{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6e0878bd-9a79-4ef1-b79a-a4604ce72746",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import os \n",
    "import time\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c012418e-0961-4ff6-8eca-436f143360fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_SNAP_ligation_07_24_2023/alignment/junc_unmapped/parn-LI_17_uni.xartifacts.xc.genome.aligned.v0.m1.unmapped.aligned.junctions.v0.m1.unmapped.fa']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#fastas = glob.glob(\"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/prg1_ligation/results_ligation/alignment/junc_unmapped/parn-LI*m1000.unmapped.fa\")\n",
    "fastas = glob.glob(\"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_SNAP_ligation_07_24_2023/alignment/junc_unmapped/parn-LI_17_uni.xartifacts.xc.genome.aligned.v0.m1.unmapped.aligned.junctions.v0.m1.unmapped.fa\")\n",
    "#fastas = glob.glob(\"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_SNAP_ligation_07_24_2023/alignment/junc_unmapped/WT-LI_17_uni.xartifacts.xc.genome.aligned.v0.m1.unmapped.aligned.junctions.v0.m1.unmapped.fa\")\n",
    "#fastas = glob.glob(\"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/prg1_ligation/results_ligation/alignment/junc_unmapped/WT-LI*m1000.unmapped.fa\")\n",
    "fastas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3809677d-ed44-42e7-80d9-432236325d52",
   "metadata": {},
   "outputs": [],
   "source": [
    "piRNA_ta_index = '/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/prg1_ligation/tailor_index/uniq_piRNA_50nt'\n",
    "piRNA_bt_index = '/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/prg1_ligation/bowtie_index/uniq_piRNA_50nt'\n",
    "genome_bt_index = '/fs/ess/PCON0160/ben/pipelines/nextflow_smRNA/index/bowtie/c_elegans.PRJNA13758.WS279.genomic/c_elegans.PRJNA13758.WS279.genomic'\n",
    "tmpdir = '/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/prg1_ligation/tmp'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "01da4791-cec1-44dd-869b-310381c4a86d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_ligation(fastas, piRNA_ta_index, piRNA_bt_index, genome_bt_index, tmpdir) : \n",
    "    \n",
    "    for e,f in enumerate(fastas) : \n",
    "        \n",
    "        sample = os.path.basename(f).split(\".\")[0]\n",
    "        print(sample)\n",
    "        \n",
    "        cmd = f\"sh /fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/prg1_ligation/tailor_map.sh {f} {piRNA_ta_index} {piRNA_bt_index} {genome_bt_index} {tmpdir}\"\n",
    "        os.system(cmd)\n",
    "        \n",
    "        candidates = pd.read_csv(f\"{tmpdir}/candidates.bed\", \n",
    "                                 names = ['chrom', 'start', 'end', 'full_seq', 'count', 'strand', 'seq_minus_unal', 'unal'],\n",
    "                                 sep = \"\\t\",\n",
    "                                 header = None\n",
    "                                )\n",
    "        \n",
    "        print(f\"Candidates with piRNA sequence {candidates['count'].sum()}\")\n",
    "        \n",
    "        aln_anti = pd.read_csv(f\"{tmpdir}/anti.bed\",\n",
    "                               names = ['chrom', 'anti_start', 'anti_end', 'seq'],\n",
    "                               sep = \"\\t\",\n",
    "                               usecols = [0,1,2,3],\n",
    "                               header = None\n",
    "                              )\n",
    "        \n",
    "        aln_anti[['full_seq','unal','x_count']] = aln_anti.seq.str.split(\":\" ,expand=True)\n",
    "        aln_anti = aln_anti.drop(columns = ['seq'])\n",
    "        \n",
    "        candidates_filt = aln_anti.merge(candidates, on = ['chrom', 'full_seq', 'unal'], how = 'inner')\n",
    "        \n",
    "        #print(candidates.shape[0])\n",
    "        #print(aln_anti.shape[0])\n",
    "        #print(candidates_filt.shape[0])\n",
    "        #print(candidates_filt.head())\n",
    "        \n",
    "        candidates_filt['sample'] = sample\n",
    "        print(f\"Candidates with piRNA AND anti sequence {candidates_filt['count'].sum()}\")\n",
    "        if e == 0 : \n",
    "            res = candidates_filt \n",
    "        else : \n",
    "            res = pd.concat([res,candidates_filt], ignore_index = True)\n",
    "        \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4b1302a-21ca-47e4-a861-a434df389efd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parn-LI_17_uni\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of starting reads = 3408789.00 \n",
      "Number of reads with tails greater than 10nt long = 383.00 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "# reads processed: 7365\n",
      "# reads with at least one alignment: 6346 (86.16%)\n",
      "# reads that failed to align: 1019 (13.84%)\n",
      "# reads with alignments suppressed due to -m: 4227 (57.39%)\n",
      "Reported 2119 alignments\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reads uniquely map to genome = 2463.00 \n",
      "Number of reads mapping antisense to a piRNA = 1871.00 \n",
      "Candidates with piRNA sequence 8698\n",
      "Candidates with piRNA AND anti sequence 1858\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "# reads processed: 2119\n",
      "# reads with at least one alignment: 1636 (77.21%)\n",
      "# reads that failed to align: 483 (22.79%)\n",
      "# reads with alignments suppressed due to -m: 1 (0.05%)\n",
      "Reported 1635 alignments\n"
     ]
    }
   ],
   "source": [
    "res = process_ligation(fastas, piRNA_ta_index, piRNA_bt_index, genome_bt_index, tmpdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5bdec3ea-eed1-4781-b277-b0e3b42f331e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>anti_start</th>\n",
       "      <th>anti_end</th>\n",
       "      <th>full_seq</th>\n",
       "      <th>unal</th>\n",
       "      <th>x_count</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>count</th>\n",
       "      <th>strand</th>\n",
       "      <th>seq_minus_unal</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>WBGene00045637,Y64G10A.19,21ur-3513</td>\n",
       "      <td>9</td>\n",
       "      <td>29</td>\n",
       "      <td>TTAGTTCTTTGCATAGTACCTTCCAAAACGTTTTGGAAGGTACTATGCA</td>\n",
       "      <td>GTTTTGGAAGGTACTATGCA</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGTTCTTTGCATAGTACCTTCCAAAAC</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>WBGene00048838,Y7A9A.17,21ur-1343</td>\n",
       "      <td>9</td>\n",
       "      <td>27</td>\n",
       "      <td>TGAAGGAAGAGTACGAAACTTTTGTAATTACAAAAGTTTCGTACT</td>\n",
       "      <td>TTACAAAAGTTTCGTACT</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>TGAAGGAAGAGTACGAAACTTTTGTAA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>WBGene00048268,Y105C5A.274,21ur-977</td>\n",
       "      <td>10</td>\n",
       "      <td>27</td>\n",
       "      <td>TTCGTGACACATCGCTAGTCCATTAACGTTAATGGACTAGCGAT</td>\n",
       "      <td>GTTAATGGACTAGCGAT</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>8</td>\n",
       "      <td>+</td>\n",
       "      <td>TTCGTGACACATCGCTAGTCCATTAAC</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>WBGene00045894,Y116A8C.60,21ur-3545</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>TCATAGAGTAAGATTTGTTCTGTCTAATATTAGACAGAACAAATCTT</td>\n",
       "      <td>ATTAGACAGAACAAATCTT</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>TCATAGAGTAAGATTTGTTCTGTCTAAT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>WBGene00048838,Y7A9A.17,21ur-1343</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>TGAAGGAAGAGTACGAAACTTTTGTAAATTTACAAAAGTTTCGTACT</td>\n",
       "      <td>TTTACAAAAGTTTCGTACT</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>+</td>\n",
       "      <td>TGAAGGAAGAGTACGAAACTTTTGTAAA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>WBGene00048186,Y67A10A.76,21ur-208</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCATCTTATGACATCTGCTTAA</td>\n",
       "      <td>CTTATGACATCTGCTTAA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCAT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>WBGene00048186,Y67A10A.76,21ur-208</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCACTTATGACATCTGCTTAA</td>\n",
       "      <td>CTTATGACATCTGCTTAA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>WBGene00048186,Y67A10A.76,21ur-208</td>\n",
       "      <td>10</td>\n",
       "      <td>27</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCATTTATGACATCTGCTTAA</td>\n",
       "      <td>TTATGACATCTGCTTAA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCAT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>WBGene00048186,Y67A10A.76,21ur-208</td>\n",
       "      <td>10</td>\n",
       "      <td>27</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCATAATTATGACATCTGCTTAA</td>\n",
       "      <td>TTATGACATCTGCTTAA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGTGTTCCTTAAGCAGATGTCATAA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1532</th>\n",
       "      <td>WBGene00175007,Y7A9D.61,21ur-5640</td>\n",
       "      <td>9</td>\n",
       "      <td>27</td>\n",
       "      <td>TACAGATCAGCTGATGAACATACTATAAGTATGTTCATCAGC</td>\n",
       "      <td>ATAAGTATGTTCATCAGC</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TACAGATCAGCTGATGAACATACT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1533 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    chrom  anti_start  anti_end  \\\n",
       "20    WBGene00045637,Y64G10A.19,21ur-3513           9        29   \n",
       "736     WBGene00048838,Y7A9A.17,21ur-1343           9        27   \n",
       "574   WBGene00048268,Y105C5A.274,21ur-977          10        27   \n",
       "89    WBGene00045894,Y116A8C.60,21ur-3545           9        28   \n",
       "737     WBGene00048838,Y7A9A.17,21ur-1343           9        28   \n",
       "...                                   ...         ...       ...   \n",
       "541    WBGene00048186,Y67A10A.76,21ur-208          10        28   \n",
       "540    WBGene00048186,Y67A10A.76,21ur-208          10        28   \n",
       "539    WBGene00048186,Y67A10A.76,21ur-208          10        27   \n",
       "538    WBGene00048186,Y67A10A.76,21ur-208          10        27   \n",
       "1532    WBGene00175007,Y7A9D.61,21ur-5640           9        27   \n",
       "\n",
       "                                               full_seq                  unal  \\\n",
       "20    TTAGTTCTTTGCATAGTACCTTCCAAAACGTTTTGGAAGGTACTATGCA  GTTTTGGAAGGTACTATGCA   \n",
       "736       TGAAGGAAGAGTACGAAACTTTTGTAATTACAAAAGTTTCGTACT    TTACAAAAGTTTCGTACT   \n",
       "574        TTCGTGACACATCGCTAGTCCATTAACGTTAATGGACTAGCGAT     GTTAATGGACTAGCGAT   \n",
       "89      TCATAGAGTAAGATTTGTTCTGTCTAATATTAGACAGAACAAATCTT   ATTAGACAGAACAAATCTT   \n",
       "737     TGAAGGAAGAGTACGAAACTTTTGTAAATTTACAAAAGTTTCGTACT   TTTACAAAAGTTTCGTACT   \n",
       "...                                                 ...                   ...   \n",
       "541         TATGTGTTCCTTAAGCAGATGTCATCTTATGACATCTGCTTAA    CTTATGACATCTGCTTAA   \n",
       "540          TATGTGTTCCTTAAGCAGATGTCACTTATGACATCTGCTTAA    CTTATGACATCTGCTTAA   \n",
       "539          TATGTGTTCCTTAAGCAGATGTCATTTATGACATCTGCTTAA     TTATGACATCTGCTTAA   \n",
       "538        TATGTGTTCCTTAAGCAGATGTCATAATTATGACATCTGCTTAA     TTATGACATCTGCTTAA   \n",
       "1532         TACAGATCAGCTGATGAACATACTATAAGTATGTTCATCAGC    ATAAGTATGTTCATCAGC   \n",
       "\n",
       "     x_count  start  end  count strand                 seq_minus_unal  \\\n",
       "20         8      0   29      8      +  TTAGTTCTTTGCATAGTACCTTCCAAAAC   \n",
       "736        8      0   27      8      +    TGAAGGAAGAGTACGAAACTTTTGTAA   \n",
       "574        8      0   27      8      +    TTCGTGACACATCGCTAGTCCATTAAC   \n",
       "89         7      0   28      7      +   TCATAGAGTAAGATTTGTTCTGTCTAAT   \n",
       "737        6      0   28      6      +   TGAAGGAAGAGTACGAAACTTTTGTAAA   \n",
       "...      ...    ...  ...    ...    ...                            ...   \n",
       "541        1      0   25      1      +      TATGTGTTCCTTAAGCAGATGTCAT   \n",
       "540        1      0   24      1      +       TATGTGTTCCTTAAGCAGATGTCA   \n",
       "539        1      0   25      1      +      TATGTGTTCCTTAAGCAGATGTCAT   \n",
       "538        1      0   27      1      +    TATGTGTTCCTTAAGCAGATGTCATAA   \n",
       "1532       1      0   24      1      +       TACAGATCAGCTGATGAACATACT   \n",
       "\n",
       "              sample  \n",
       "20    parn-LI_17_uni  \n",
       "736   parn-LI_17_uni  \n",
       "574   parn-LI_17_uni  \n",
       "89    parn-LI_17_uni  \n",
       "737   parn-LI_17_uni  \n",
       "...              ...  \n",
       "541   parn-LI_17_uni  \n",
       "540   parn-LI_17_uni  \n",
       "539   parn-LI_17_uni  \n",
       "538   parn-LI_17_uni  \n",
       "1532  parn-LI_17_uni  \n",
       "\n",
       "[1533 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.sort_values(by = ['count'], ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "489f820b-e349-48b1-bcd0-02b546d2d735",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>parn-LI_17_uni</th>\n",
       "      <td>1533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                chrom\n",
       "sample               \n",
       "parn-LI_17_uni   1533"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.groupby(['sample']).agg({'chrom':'size'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8147ecf1-b407-4f28-99e0-3d05ea6a8e35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>parn-LI_17_uni</th>\n",
       "      <td>1858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                count\n",
       "sample               \n",
       "parn-LI_17_uni   1858"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.groupby(['sample']).agg({'count':'sum'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3a0f59a9-51de-4d39-a137-1015ed4fcaf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>anti_start</th>\n",
       "      <th>anti_end</th>\n",
       "      <th>full_seq</th>\n",
       "      <th>unal</th>\n",
       "      <th>x_count</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>count</th>\n",
       "      <th>strand</th>\n",
       "      <th>seq_minus_unal</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00045523,M199.10,21ur-4900</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>TTGAGCATCAATTATCATTCTAATATGAATTAGAATGATAATT</td>\n",
       "      <td>ATGAATTAGAATGATAATT</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TTGAGCATCAATTATCATTCTAAT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00045531,Y105C5A.35,21ur-3262</td>\n",
       "      <td>8</td>\n",
       "      <td>28</td>\n",
       "      <td>TCGAGGATTAATCGAGAATGAATTTTGAAAAATTCATTCTCGATTA</td>\n",
       "      <td>GAAAAATTCATTCTCGATTA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGAGGATTAATCGAGAATGAATTTT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00045532,C02B10.7,21ur-116</td>\n",
       "      <td>9</td>\n",
       "      <td>25</td>\n",
       "      <td>TGGGGTTGTAGATATAGAAGACTTATATAAGTCTTCTATATCT</td>\n",
       "      <td>TAAGTCTTCTATATCT</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TGGGGTTGTAGATATAGAAGACTTATA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00045538,Y51H4A.36,21ur-783</td>\n",
       "      <td>10</td>\n",
       "      <td>26</td>\n",
       "      <td>TAGGACACATCCCAGCCTGTCTTAGTAAGACAGGCTGGGA</td>\n",
       "      <td>GTAAGACAGGCTGGGA</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>TAGGACACATCCCAGCCTGTCTTA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00045541,C35D6.15,21ur-2515</td>\n",
       "      <td>13</td>\n",
       "      <td>29</td>\n",
       "      <td>TCTCATATGGTAACTTTATTTCCACGTTGTAAACAACGTGGAAATAAAG</td>\n",
       "      <td>CAACGTGGAAATAAAG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TCTCATATGGTAACTTTATTTCCACGTTGTAAA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>WBGene00174958,Y17G9A.84,21ur-9338</td>\n",
       "      <td>9</td>\n",
       "      <td>27</td>\n",
       "      <td>TAGGTACTGAACTTGTCTGAATTTATGAAATTCAGACAAGTT</td>\n",
       "      <td>ATGAAATTCAGACAAGTT</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TAGGTACTGAACTTGTCTGAATTT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529</th>\n",
       "      <td>WBGene00174964,Y73F8A.1157,21ur-7551</td>\n",
       "      <td>9</td>\n",
       "      <td>27</td>\n",
       "      <td>TGGTTCCTTCCTTCCGCCTTTGGTATATATATACCAAAGGCGGAAGG</td>\n",
       "      <td>TATACCAAAGGCGGAAGG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TGGTTCCTTCCTTCCGCCTTTGGTATATA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1530</th>\n",
       "      <td>WBGene00174964,Y73F8A.1157,21ur-7551</td>\n",
       "      <td>10</td>\n",
       "      <td>23</td>\n",
       "      <td>TGGTTCCTTCCTTCCGCCTTTGGTATATACCAAAGGCGGAAG</td>\n",
       "      <td>CCAAAGGCGGAAG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TGGTTCCTTCCTTCCGCCTTTGGTATATA</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1531</th>\n",
       "      <td>WBGene00174989,C47A4.140,21ur-8457</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>TCTTTCTTTGTAACCCATTAATCATTGTCAACAATGATTAATGGGT...</td>\n",
       "      <td>CAACAATGATTAATGGGTTACA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TCTTTCTTTGTAACCCATTAATCATTGT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1532</th>\n",
       "      <td>WBGene00175007,Y7A9D.61,21ur-5640</td>\n",
       "      <td>9</td>\n",
       "      <td>27</td>\n",
       "      <td>TACAGATCAGCTGATGAACATACTATAAGTATGTTCATCAGC</td>\n",
       "      <td>ATAAGTATGTTCATCAGC</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>TACAGATCAGCTGATGAACATACT</td>\n",
       "      <td>parn-LI_17_uni</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1533 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     chrom  anti_start  anti_end  \\\n",
       "0         WBGene00045523,M199.10,21ur-4900           9        28   \n",
       "1      WBGene00045531,Y105C5A.35,21ur-3262           8        28   \n",
       "2         WBGene00045532,C02B10.7,21ur-116           9        25   \n",
       "3        WBGene00045538,Y51H4A.36,21ur-783          10        26   \n",
       "4        WBGene00045541,C35D6.15,21ur-2515          13        29   \n",
       "...                                    ...         ...       ...   \n",
       "1528    WBGene00174958,Y17G9A.84,21ur-9338           9        27   \n",
       "1529  WBGene00174964,Y73F8A.1157,21ur-7551           9        27   \n",
       "1530  WBGene00174964,Y73F8A.1157,21ur-7551          10        23   \n",
       "1531    WBGene00174989,C47A4.140,21ur-8457           8        30   \n",
       "1532     WBGene00175007,Y7A9D.61,21ur-5640           9        27   \n",
       "\n",
       "                                               full_seq  \\\n",
       "0           TTGAGCATCAATTATCATTCTAATATGAATTAGAATGATAATT   \n",
       "1        TCGAGGATTAATCGAGAATGAATTTTGAAAAATTCATTCTCGATTA   \n",
       "2           TGGGGTTGTAGATATAGAAGACTTATATAAGTCTTCTATATCT   \n",
       "3              TAGGACACATCCCAGCCTGTCTTAGTAAGACAGGCTGGGA   \n",
       "4     TCTCATATGGTAACTTTATTTCCACGTTGTAAACAACGTGGAAATAAAG   \n",
       "...                                                 ...   \n",
       "1528         TAGGTACTGAACTTGTCTGAATTTATGAAATTCAGACAAGTT   \n",
       "1529    TGGTTCCTTCCTTCCGCCTTTGGTATATATATACCAAAGGCGGAAGG   \n",
       "1530         TGGTTCCTTCCTTCCGCCTTTGGTATATACCAAAGGCGGAAG   \n",
       "1531  TCTTTCTTTGTAACCCATTAATCATTGTCAACAATGATTAATGGGT...   \n",
       "1532         TACAGATCAGCTGATGAACATACTATAAGTATGTTCATCAGC   \n",
       "\n",
       "                        unal x_count  start  end  count strand  \\\n",
       "0        ATGAATTAGAATGATAATT       1      0   24      1      +   \n",
       "1       GAAAAATTCATTCTCGATTA       1      0   26      1      +   \n",
       "2           TAAGTCTTCTATATCT       1      0   27      1      +   \n",
       "3           GTAAGACAGGCTGGGA       2      1   25      2      +   \n",
       "4           CAACGTGGAAATAAAG       1      0   33      1      +   \n",
       "...                      ...     ...    ...  ...    ...    ...   \n",
       "1528      ATGAAATTCAGACAAGTT       1      0   24      1      +   \n",
       "1529      TATACCAAAGGCGGAAGG       1      0   29      1      +   \n",
       "1530           CCAAAGGCGGAAG       1      0   29      1      +   \n",
       "1531  CAACAATGATTAATGGGTTACA       1      0   28      1      +   \n",
       "1532      ATAAGTATGTTCATCAGC       1      0   24      1      +   \n",
       "\n",
       "                         seq_minus_unal          sample  \n",
       "0              TTGAGCATCAATTATCATTCTAAT  parn-LI_17_uni  \n",
       "1            TCGAGGATTAATCGAGAATGAATTTT  parn-LI_17_uni  \n",
       "2           TGGGGTTGTAGATATAGAAGACTTATA  parn-LI_17_uni  \n",
       "3              TAGGACACATCCCAGCCTGTCTTA  parn-LI_17_uni  \n",
       "4     TCTCATATGGTAACTTTATTTCCACGTTGTAAA  parn-LI_17_uni  \n",
       "...                                 ...             ...  \n",
       "1528           TAGGTACTGAACTTGTCTGAATTT  parn-LI_17_uni  \n",
       "1529      TGGTTCCTTCCTTCCGCCTTTGGTATATA  parn-LI_17_uni  \n",
       "1530      TGGTTCCTTCCTTCCGCCTTTGGTATATA  parn-LI_17_uni  \n",
       "1531       TCTTTCTTTGTAACCCATTAATCATTGT  parn-LI_17_uni  \n",
       "1532           TACAGATCAGCTGATGAACATACT  parn-LI_17_uni  \n",
       "\n",
       "[1533 rows x 12 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b0eae75d-cd78-48ac-8e50-807f1b53ce9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_filt = res.query(\n",
    "    'seq_minus_unal.str.len() == 21'\n",
    ").reset_index(drop = True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "bfb28f4c-b365-45b2-9676-d3f55b6e5dbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>anti_start</th>\n",
       "      <th>anti_end</th>\n",
       "      <th>full_seq</th>\n",
       "      <th>unal</th>\n",
       "      <th>x_count</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>count</th>\n",
       "      <th>strand</th>\n",
       "      <th>seq_minus_unal</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, anti_start, anti_end, full_seq, unal, x_count, start, end, count, strand, seq_minus_unal, sample]\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "16bf10db-632e-47bb-a9ee-007478b5166f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>parn-LI_17_uni</th>\n",
       "      <td>1858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                count\n",
       "sample               \n",
       "parn-LI_17_uni   1858"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.groupby(['sample']).agg({'count':'sum'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "685b64a7-1c68-4c37-be5a-57ec847048bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_filt.to_csv(\"prg1_ligation.tsv\", sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d174fbc3-ed13-407d-b5d4-229c96905de7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>anti_start</th>\n",
       "      <th>anti_end</th>\n",
       "      <th>full_seq</th>\n",
       "      <th>unal</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>count</th>\n",
       "      <th>strand</th>\n",
       "      <th>seq_minus_unal</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, anti_start, anti_end, full_seq, unal, start, end, count, strand, seq_minus_unal, sample]\n",
       "Index: []"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cca01d53-f3ba-4fc9-b1fd-438c502393bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_filt.groupby(['chrom', 'sample']).agg( count = ('count', 'sum')).reset_index().to_csv(\"parn1_ligation_counts.tsv\", sep = \"\\t\", header = True, index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9 (Conda 2022.05) [python/3.9-2022.05]",
   "language": "python",
   "name": "python39_202205"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
