{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b47a17ee-5998-4c74-bbf0-57825cc85b19",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import os \n",
    "import time\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b62172c-e93f-4a62-b5c5-cce8b143573d",
   "metadata": {},
   "source": [
    "# Filter piRNA loci / filter antisense mapping reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6e8b6a1d-b968-4976-945b-e733eb9e4c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first make bed with uni piRNA loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4ee5bdb7-1985-4bb2-b424-a6fab618f876",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7206 c_elegans.PRJNA13758.WS279.distinct_piRNA_loci.bed\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# one way to do this is with bedtools merge\n",
    "\n",
    "module load python\n",
    "source activate rnaseq_basic\n",
    "\n",
    "reference=/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/c_elegans.PRJNA13758.WS279.slim.bed\n",
    "\n",
    "cat $reference |\\\n",
    "    #awk '($5==\"gene\")' |\\\n",
    "    #grep piRNA |\\\n",
    "    grep exon |\\\n",
    "    #egrep \"piRNA|miRNA|protein_coding|pseudogene\" |\\\n",
    "    awk -F'\\t' '{OFS=\"\\t\"; print $1,$2-10,$3+10,$7,$8,$6}' |\\\n",
    "    awk -F'\\t' '{OFS=\"\\t\"; if ($2<0) print $1,0,$3,$4,$5,$6; else print $0}' |\\\n",
    "    sort -k1,1 -k2,2n |\\\n",
    "    bedtools merge -c 1,4,5,6 -o count,collapse,collapse,collapse |\\\n",
    "    awk '($4==1)' |\\\n",
    "    grep piRNA |\\\n",
    "    awk -F'\\t' '{OFS=\"\\t\"; if ($7==\"+\") print $1,$2+10,$3,$5,$6,$7; else print $1,$2,$3-10,$5,$6,$7}' \\\n",
    "    > c_elegans.PRJNA13758.WS279.distinct_piRNA_loci.bed\n",
    "\n",
    "wc -l c_elegans.PRJNA13758.WS279.distinct_piRNA_loci.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "27047511-d736-436f-9c7a-adf6c36de278",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uni_piRNA_loci(bed) : \n",
    "    \n",
    "    \"\"\"\n",
    "    Make unique piRNA locus bedfile\n",
    "    \"\"\"\n",
    "    \n",
    "    chroms, starts, ends, genes, strands, biotypes = [], [], [], [], [], []\n",
    "    with open(bed, 'r') as f : \n",
    "        for line in f : \n",
    "            info = line.strip().split(\"\\t\")\n",
    "            chrom = str(info[0])\n",
    "            feature = str(info[4])\n",
    "            start = int(info[1])\n",
    "            end = int(info[2])\n",
    "            strand = str(info[5])\n",
    "            gene = str(info[6])\n",
    "            biotype = str(info[7])\n",
    "            \n",
    "            #if (feature == \"gene\" and biotype == \"piRNA\") or (feature == \"gene\" and biotype == \"redefine_type2_piRNA\") or (feature == \"gene\" and biotype == \"miRNA\") or (feature == \"exon\" and biotype == \"protein_coding\") or (feature == \"exon\" and biotype == \"pseudogene\") or (biotype == \"lincRNA\") or (biotype == \"ncRNA\") :\n",
    "            if (feature == \"gene\" and biotype == \"piRNA\") or (feature == \"gene\" and biotype == \"miRNA\") or (feature == \"exon\" and biotype == \"protein_coding\") or (feature == \"exon\" and biotype == \"pseudogene\") or (biotype == \"lincRNA\") or (biotype == \"ncRNA\") :\n",
    "                chroms.append(chrom)\n",
    "                starts.append(start)\n",
    "                ends.append(end)\n",
    "                genes.append(gene)\n",
    "                strands.append(strand)\n",
    "                biotypes.append(biotype)\n",
    "    f.close() \n",
    "    \n",
    "    df = pd.DataFrame({\n",
    "        'chrom':chroms,\n",
    "        'start':starts,\n",
    "        'end':ends,\n",
    "        'gene':genes,\n",
    "        'strand':strands,\n",
    "        'biotype':biotypes\n",
    "    })\n",
    "    \n",
    "    df['start'] = df['start']\n",
    "    df['end'] = df['end']\n",
    "    sort = df.sort_values(by = ['chrom', 'start'], ascending = True).reset_index(drop = True)\n",
    "    \n",
    "    sort.to_csv(\"sorted_bed.tsv\", sep = \"\\t\", header = True, index = False)\n",
    "    idx = []\n",
    "    \n",
    "    #\n",
    "    #   U------------>\n",
    "    #      U-------------> \n",
    "    #          \n",
    "    k = 0\n",
    "    for i,row in sort.iterrows() :\n",
    "        if i < len(sort)-1 :\n",
    "            \n",
    "            # check nearest right neighbor\n",
    "            if int(sort.iloc[i+1]['start'])-10 > int(sort.iloc[i]['end'])+10 :\n",
    "                \n",
    "                # check nearest left neighbor\n",
    "                if int(sort.iloc[i]['start'])-10 > int(sort.iloc[i-1]['end'])+10 :\n",
    "                    \n",
    "                    idx.append(i)\n",
    "    \n",
    "    uni = sort.iloc[idx]\n",
    "    uni = uni.reset_index(drop = True)\n",
    "    uni_piRNA = uni.query('biotype == \"piRNA\"').reset_index(drop = True).drop(columns = ['biotype'])\n",
    "    return uni_piRNA\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6726816a-439b-45b0-b3ba-6ae78a64df09",
   "metadata": {},
   "outputs": [],
   "source": [
    "uni = uni_piRNA_loci(\"/fs/ess/PCON0160/ben/genomes/c_elegans/WS279/c_elegans.PRJNA13758.WS279.slim.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ebe56984-ae91-4d77-b126-1d858fd9bfcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "uni.to_csv(\"uniq_piRNA.tsv\", sep = \"\\t\", header = False, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0a383a25-31f4-4ae2-80eb-50ac1c991663",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7318"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uni.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "dc5befed-1748-4b52-9961-728aefccdda7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bed(bedfile) : \n",
    "    \n",
    "    d = pd.read_csv(bedfile, usecols = [0, 1, 2, 3, 7, 9, 10, 13], header = 0, names = ['chrom', 'start', 'end', 'seq', 'biotype', 'feature', 'gene', 'count'], sep = \"\\t\")\n",
    "    #d = pd.read_csv(bedfile, usecols = [0,1,2,3,9,10,14], header = 0, names = ['chrom', 'start', 'end', 'seq', 'feature', 'gene', 'count'], sep = \"\\t\")\n",
    "\n",
    "    sub = d.query('feature == \"antipiRNA\"').reset_index(drop = True)\n",
    "\n",
    "    return sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "aa39b4c8-9390-45b9-b03d-207634263535",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_piRNA(df) : \n",
    "        \n",
    "    antipiRNA = df.query(\n",
    "        'feature == \"antipiRNA\" and (seq.str[0] == \"G\" or seq.str[0]==\"A\") and (seq.str.len()>=17 and seq.str.len()<=19)'\n",
    "    ).reset_index(\n",
    "        drop = True\n",
    "    )\n",
    "    \n",
    "    sense = antipiRNA.query('strand == \"+\"').reset_index(drop = True)\n",
    "    anti = antipiRNA.query('strand == \"-\"').reset_index(drop = True)\n",
    "    \n",
    "    # for antisense piRNA further filter by removing situations like this\n",
    "    #           rs U-----------------> re\n",
    "    #          s <----------G/A e\n",
    "\n",
    "    #          rs <-----------------U re\n",
    "    #                   s G/A------------> e\n",
    "    \n",
    "    filt_sense = sense.query('ref_start < start').reset_index(drop = True)\n",
    "    \n",
    "    filt_anti = anti.query('end < ref_end').reset_index(drop = True)\n",
    "    \n",
    "    anti_piRNA_filt = pd.concat([filt_sense, filt_anti], ignore_index = True)\n",
    "    \n",
    "    return anti_piRNA_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "a07b4369-fa9d-4b29-b6d7-13e162cd9368",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_piRNA(uniq_piRNA, BEDFILES) :\n",
    "    \n",
    "    uni = pd.read_csv(uniq_piRNA, names = ['chrom', 'start', 'end', 'gene','strand'], usecols = [0,1,2,3,4], sep = \"\\t\")\n",
    "    uni_piRNA = uni['gene'].tolist()\n",
    "    \n",
    "    uni = uni.rename(columns = {'start':'ref_start', 'end':'ref_end'})\n",
    "\n",
    "    chroms = []\n",
    "    starts = []\n",
    "    ends = []\n",
    "    seqs = []\n",
    "    counts = []\n",
    "    strands = []\n",
    "    genes = []\n",
    "    features = []\n",
    "    for e,B in enumerate(BEDFILES) : \n",
    "        print(os.path.basename(B))\n",
    "        d = read_bed(B)\n",
    "        d['sample'] = os.path.basename(B).split(\".\")[0]\n",
    "\n",
    "        d = d.merge(uni, on = ['gene', 'chrom'], how = 'inner')\n",
    "        \n",
    "        if e == 0 : \n",
    "            table = d\n",
    "        else : \n",
    "            table = pd.concat([table, d])\n",
    "            \n",
    "    filt = filter_piRNA(table)\n",
    "                \n",
    "    return filt\n",
    "             "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "510a04dd-ea90-4752-8d13-843d9d558d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DIRS = ['/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_07102023/bed', \n",
    "#        '/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_07_14_2023/bed',\n",
    "#        '/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_oxi_07_18_2023/bed',\n",
    "#        '/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_SNAP_ligation_07_24_2023/bed']\n",
    "#DIRS = ['/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_normalize_miRNA_07_31_2023/bed']\n",
    "#DIRS = ['/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_07_14_2023/bed']\n",
    "#DIRS = ['/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_type2_piRNA_08_20_2023/bed']\n",
    "\n",
    "DIRS = ['/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_07102023/bed']\n",
    "\n",
    "BEDFILES = []\n",
    "\n",
    "for D in DIRS : \n",
    "    bed = glob.glob(f\"{D}/*parn_*YD*.bed.tsv\")\n",
    "    BEDFILES += bed\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "996aeded-fe54-4ff6-80e2-adb8706e173d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "parn_YD2.trimmed.uniq.xartifacts.xc.aligned.v0.m1.bed.tsv\n",
      "parn_YD1.trimmed.uniq.xartifacts.xc.aligned.v0.m1.bed.tsv\n"
     ]
    }
   ],
   "source": [
    "outname = \"antipiRNA_09222023\"\n",
    "\n",
    "filt = process_piRNA(uniq_piRNA = \"./uniq_piRNA.tsv\",\n",
    "                  BEDFILES = BEDFILES)\n",
    "\n",
    "filt.to_csv(f\"./{outname}.tsv\", sep = \"\\t\", header = True, index = False)\n",
    "#else : \n",
    "#    filt = pd.read_csv(f\"./{outname}.tsv\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "5b75a7f5-c5f6-48e4-9ce6-62b941974053",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>seq</th>\n",
       "      <th>biotype</th>\n",
       "      <th>feature</th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "      <th>sample</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1711</th>\n",
       "      <td>IV</td>\n",
       "      <td>6050117</td>\n",
       "      <td>6050135</td>\n",
       "      <td>GTGGAATTTGAGGAAACA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00049633,H16O14.17,21ur-2732</td>\n",
       "      <td>7.181639</td>\n",
       "      <td>parn_YD1</td>\n",
       "      <td>6050108</td>\n",
       "      <td>6050129</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3680</th>\n",
       "      <td>IV</td>\n",
       "      <td>14697883</td>\n",
       "      <td>14697902</td>\n",
       "      <td>AGATAGTGGGATTAGAATA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00169779,Y57G11C.797,21ur-8259</td>\n",
       "      <td>6.320672</td>\n",
       "      <td>parn_YD2</td>\n",
       "      <td>14697890</td>\n",
       "      <td>14697911</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>IV</td>\n",
       "      <td>6050117</td>\n",
       "      <td>6050135</td>\n",
       "      <td>GTGGAATTTGAGGAAACA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00049633,H16O14.17,21ur-2732</td>\n",
       "      <td>5.980328</td>\n",
       "      <td>parn_YD2</td>\n",
       "      <td>6050108</td>\n",
       "      <td>6050129</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5160</th>\n",
       "      <td>IV</td>\n",
       "      <td>14697883</td>\n",
       "      <td>14697902</td>\n",
       "      <td>AGATAGTGGGATTAGAATA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00169779,Y57G11C.797,21ur-8259</td>\n",
       "      <td>4.899103</td>\n",
       "      <td>parn_YD1</td>\n",
       "      <td>14697890</td>\n",
       "      <td>14697911</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5056</th>\n",
       "      <td>IV</td>\n",
       "      <td>14580095</td>\n",
       "      <td>14580114</td>\n",
       "      <td>ACGAGTTTGCTTCACTAGA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00048342,Y57G11B.57,21ur-635</td>\n",
       "      <td>4.843431</td>\n",
       "      <td>parn_YD1</td>\n",
       "      <td>14580101</td>\n",
       "      <td>14580122</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3133</th>\n",
       "      <td>IV</td>\n",
       "      <td>5970851</td>\n",
       "      <td>5970870</td>\n",
       "      <td>AGTATGGGGTTCATTAAAT</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00048889,B0350.20,21ur-3556</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>parn_YD2</td>\n",
       "      <td>5970858</td>\n",
       "      <td>5970879</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3135</th>\n",
       "      <td>IV</td>\n",
       "      <td>5981461</td>\n",
       "      <td>5981480</td>\n",
       "      <td>AATGGGGCTTGATGGCCGT</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00047205,B0350.11,21ur-1749</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>parn_YD2</td>\n",
       "      <td>5981471</td>\n",
       "      <td>5981492</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3136</th>\n",
       "      <td>IV</td>\n",
       "      <td>5981462</td>\n",
       "      <td>5981479</td>\n",
       "      <td>ATGGGGCTTGATGGCCG</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00047205,B0350.11,21ur-1749</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>parn_YD2</td>\n",
       "      <td>5981471</td>\n",
       "      <td>5981492</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3140</th>\n",
       "      <td>IV</td>\n",
       "      <td>6034565</td>\n",
       "      <td>6034584</td>\n",
       "      <td>ATATATATTTCATCTCGAA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00166051,C46G7.42,21ur-6086</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>parn_YD2</td>\n",
       "      <td>6034571</td>\n",
       "      <td>6034592</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>II</td>\n",
       "      <td>5308111</td>\n",
       "      <td>5308128</td>\n",
       "      <td>ATGAAAACGCGCTAAGT</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00174410,F08B1.6,21ur-10456</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>parn_YD2</td>\n",
       "      <td>5308102</td>\n",
       "      <td>5308123</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6015 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end                  seq biotype    feature  \\\n",
       "1711    IV   6050117   6050135   GTGGAATTTGAGGAAACA   piRNA  antipiRNA   \n",
       "3680    IV  14697883  14697902  AGATAGTGGGATTAGAATA   piRNA  antipiRNA   \n",
       "196     IV   6050117   6050135   GTGGAATTTGAGGAAACA   piRNA  antipiRNA   \n",
       "5160    IV  14697883  14697902  AGATAGTGGGATTAGAATA   piRNA  antipiRNA   \n",
       "5056    IV  14580095  14580114  ACGAGTTTGCTTCACTAGA   piRNA  antipiRNA   \n",
       "...    ...       ...       ...                  ...     ...        ...   \n",
       "3133    IV   5970851   5970870  AGTATGGGGTTCATTAAAT   piRNA  antipiRNA   \n",
       "3135    IV   5981461   5981480  AATGGGGCTTGATGGCCGT   piRNA  antipiRNA   \n",
       "3136    IV   5981462   5981479    ATGGGGCTTGATGGCCG   piRNA  antipiRNA   \n",
       "3140    IV   6034565   6034584  ATATATATTTCATCTCGAA   piRNA  antipiRNA   \n",
       "0       II   5308111   5308128    ATGAAAACGCGCTAAGT   piRNA  antipiRNA   \n",
       "\n",
       "                                      gene     count    sample  ref_start  \\\n",
       "1711    WBGene00049633,H16O14.17,21ur-2732  7.181639  parn_YD1    6050108   \n",
       "3680  WBGene00169779,Y57G11C.797,21ur-8259  6.320672  parn_YD2   14697890   \n",
       "196     WBGene00049633,H16O14.17,21ur-2732  5.980328  parn_YD2    6050108   \n",
       "5160  WBGene00169779,Y57G11C.797,21ur-8259  4.899103  parn_YD1   14697890   \n",
       "5056    WBGene00048342,Y57G11B.57,21ur-635  4.843431  parn_YD1   14580101   \n",
       "...                                    ...       ...       ...        ...   \n",
       "3133     WBGene00048889,B0350.20,21ur-3556  0.048621  parn_YD2    5970858   \n",
       "3135     WBGene00047205,B0350.11,21ur-1749  0.048621  parn_YD2    5981471   \n",
       "3136     WBGene00047205,B0350.11,21ur-1749  0.048621  parn_YD2    5981471   \n",
       "3140     WBGene00166051,C46G7.42,21ur-6086  0.048621  parn_YD2    6034571   \n",
       "0        WBGene00174410,F08B1.6,21ur-10456  0.048621  parn_YD2    5308102   \n",
       "\n",
       "       ref_end strand  \n",
       "1711   6050129      +  \n",
       "3680  14697911      -  \n",
       "196    6050129      +  \n",
       "5160  14697911      -  \n",
       "5056  14580122      -  \n",
       "...        ...    ...  \n",
       "3133   5970879      -  \n",
       "3135   5981492      -  \n",
       "3136   5981492      -  \n",
       "3140   6034592      -  \n",
       "0      5308123      +  \n",
       "\n",
       "[6015 rows x 12 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt.query('feature == \"antipiRNA\"').sort_values(by = ['count'], ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "5204c631-f270-4466-9015-91928ee319b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample\n",
       "parn_YD1    1455\n",
       "parn_YD2    1511\n",
       "Name: gene, dtype: int64"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt[['sample', 'gene']].drop_duplicates().groupby(['sample'])['gene'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "47ff2741-ccd3-4228-9db1-ced096820c37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1934, 1)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt[['sample', 'gene']].query('sample == \"parn_YD1\" or sample == \"parn_YD2\"').drop(columns = ['sample']).drop_duplicates(['gene']).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dad82c5-6014-4dd2-8acf-4b9fc958f835",
   "metadata": {},
   "source": [
    "# 1. Measure 5' to 5' and 5' to 3' distance relative to 5' U"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2f10e044-6c17-4f1a-8135-a8a2da815f0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def measure_5p3p_distance(df) : \n",
    "    \n",
    "    #  5' rs U---------------------> re\n",
    "    #         3' s <-------------G/A e 5' \n",
    "    \n",
    "    #                 rs <-----------------U re 5'\n",
    "    #    5'  s G/A------------------> e\n",
    "    \n",
    "    df = df.dropna().reset_index(drop = True)\n",
    "    \n",
    "    df['5p5p'] = df.apply(lambda row: row['end'] - row['ref_start'] if row['strand'] == \"+\" else row['ref_end'] - row['start'], axis = 1 )\n",
    "    \n",
    "    df['5p3p'] = df.apply(lambda row: row['start'] - row['ref_start'] if row['strand'] == \"+\" else row['ref_end'] - row['end'], axis = 1 )\n",
    "    \n",
    "    p5p5 = df.groupby(['5p5p', 'sample']).agg( { 'count':'sum' } ).reset_index()\n",
    "    p5p5 = p5p5.rename(columns = {'5p5p' : 'distance'})\n",
    "    p5p5['type'] = '5p5p'\n",
    "    \n",
    "    p5p3 = df.groupby(['5p3p', 'sample']).agg( { 'count':'sum' } ).reset_index()\n",
    "    p5p3 = p5p3.rename(columns = {'5p3p' : 'distance'})\n",
    "    p5p3['type'] = '5p3p'\n",
    "    \n",
    "    res = pd.concat([p5p5,p5p3])\n",
    "    \n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "25513e5e-e249-4687-941a-37e67c2b250a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dists = measure_5p3p_distance(filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4602ddb-d1c1-46ec-8e32-d5a4a2b2067e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dists.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7c929567-91d4-41ad-9f45-d80a9591832f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dists.to_csv(\"antisense_to_sense_distances.tsv\", sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "46d2daa9-bfaa-4cce-adef-797cd71bdfe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find 5p3p == 8 in parn-1 SG for specific example\n",
    "df = filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e0a5899a-4625-4f82-ab0c-ce73c623c636",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['dist5p3p'] = df.apply(lambda row: row['start'] - row['ref_start'] if row['strand'] == \"+\" else row['ref_end'] - row['end'], axis = 1 )\n",
    "df.head()\n",
    "df.query('sample.str.contains(\"parn1_prg1sg\") and dist5p3p == 8').sort_values(by = ['count'], ascending = False).to_csv(\"prg1SGparn1.tsv\", sep = \"\\t\", header = True, index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "800ab8ce-303b-4a3d-986d-598a3fd8ddfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['count', 'sample', 'dist5p3p', 'gene']].groupby(\n",
    "    ['sample', 'gene', 'dist5p3p']\n",
    ").agg(\n",
    "    {'count' : 'sum'}\n",
    ").reset_index().pivot(\n",
    "    index = ['dist5p3p', 'gene'], columns = 'sample', values = 'count'\n",
    ").reset_index().fillna(\n",
    "    0\n",
    ").to_csv(\n",
    "    \"distance_5p3p.tsv\", sep = \"\\t\", header = True, index = False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ae81c75-1323-4e72-bba5-68f670c09a04",
   "metadata": {},
   "source": [
    "# 2. Export counts of antisense for each piRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e8fd3a31-a2d4-428b-9b95-a0c6388beaac",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = filt.groupby(['gene', 'feature', 'sample']).agg({ 'count':'sum' }).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee72426-50ee-4357-a2d3-5ddb9f9e7dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "184ec10d-06f9-470b-98d5-197940a3f479",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts.groupby(['sample','feature']).agg({ 'count':'sum', 'gene':'size'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ca3d40bd-c38c-41d1-80a2-0fe8480a9522",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts.to_csv(f\"./{outname}.counts.tsv\", sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df25d39e-ad99-4454-a59f-971f6ab399b8",
   "metadata": {},
   "source": [
    "# 3. Iterate through filtered bed files and make bigwigs for visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9493189c-86d9-4187-ae93-a9b6b9f90847",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# iterate through bedfiles from pipeline output, isolate piRNA mapping reads, make bigwigs\n",
    "def make_bw(direc, bwdir, chrom_sizes, feat) : \n",
    "    \n",
    "    if not os.path.exists(bwdir) : \n",
    "        os.mkdir(bwdir)\n",
    "        \n",
    "    files = glob.glob(f\"{direc}/*.bed.tsv\")\n",
    "    \n",
    "    for f in files :\n",
    "        name = os.path.basename(f).split(\".\")[0]\n",
    "        d = pd.read_csv(f, sep = \"\\t\")\n",
    "        d_sub = d.query('feature == @feat').reset_index(drop = True)\n",
    "        reformat = d_sub[['chrom', 'start', 'end', 'seq', 'count_total_norm']]\n",
    "        reformat.to_csv(\"tmp.bed\", sep = \"\\t\", header = False, index = False)\n",
    "\n",
    "        os.system(f\"bash ./bed2bw.sh tmp.bed {bwdir}/{name}.{feat}.bw {chrom_sizes} {feat}\")\n",
    "        \n",
    "        os.system(\"rm tmp.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d081352c-5443-4dc6-bbfb-1ea719f27042",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_sizes='/fs/ess/PCON0160/ben/pipelines/nextflow_smRNA/index/bowtie/c_elegans.PRJNA13758.WS279.genomic/c_elegans.PRJNA13758.WS279.genomic_chrom_sizes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c3a81c-f8f8-4b38-ad49-43cbdacab893",
   "metadata": {},
   "outputs": [],
   "source": [
    "make_bw(\"./results_07102023/bed\", \"./bw22G\", chrom_sizes, \"siRNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "4da2626f-cf6e-4cdc-8a64-7fd6955e31c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_sizes='/fs/ess/PCON0160/ben/pipelines/nextflow_smRNA/index/bowtie/c_elegans.PRJNA13758.WS279.genomic/c_elegans.PRJNA13758.WS279.genomic_chrom_sizes'\n",
    "bwdir=f'/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_type2_piRNA_08_20_2023/antipiRNA_type2'\n",
    "\n",
    "def makebws(df, bwdir, chrom_sizes) : \n",
    "    \n",
    "    if not os.path.exists(bwdir) : \n",
    "        os.mkdir(bwdir)\n",
    "\n",
    "\n",
    "    samples = df['sample'].drop_duplicates().tolist() \n",
    "\n",
    "    for i,g in enumerate(samples) : \n",
    "\n",
    "        sub = df.query('sample == @g')\n",
    "        feat = sub['feature'].drop_duplicates().tolist()[0]\n",
    "        name = os.path.basename(g).split(\".\")[0]\n",
    "        reformat = sub[['chrom', 'start', 'end', 'seq', 'count']]\n",
    "        reformat.to_csv(\"tmp.bed\", sep = \"\\t\", header = False, index = False)\n",
    "\n",
    "        os.system(f\"bash ./bed2bw.sh tmp.bed {bwdir}/{name}.{feat}.bw {chrom_sizes} {feat}\")\n",
    "        \n",
    "        os.system(\"rm tmp.bed\")\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "75353b1c-5d06-4ae8-b3c1-5e054a391010",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(f\"./antipiRNA_bed_master_filtered_type2.tsv\", sep = \"\\t\")\n",
    "makebws(df, bwdir, chrom_sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "0fee32ae-97dc-49b0-94a3-13ce335c42ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate through bedfiles from pipeline output, isolate piRNA mapping reads, make bigwigs\n",
    "def make_sense_bw(direc, bwdir, chrom_sizes) : \n",
    "    \n",
    "    if not os.path.exists(bwdir) : \n",
    "        os.mkdir(bwdir)\n",
    "        \n",
    "    files = glob.glob(f\"{direc}/*.bed.tsv\")\n",
    "    \n",
    "    for f in files :\n",
    "        name = os.path.basename(f).split(\".\")[0]\n",
    "        d = pd.read_csv(f, sep = \"\\t\")\n",
    "        d_sub = d.query('feature == \"piRNA\" and biotype == \"redefine_type2_piRNA\"').reset_index(drop = True)\n",
    "        reformat = d_sub[['chrom', 'start', 'end', 'seq', 'count_total_norm']]\n",
    "        reformat.to_csv(\"tmp.bed\", sep = \"\\t\", header = False, index = False)\n",
    "\n",
    "        os.system(f\"bash ./bed2bw.sh tmp.bed {bwdir}/{name}.piRNA.bw {chrom_sizes} piRNA\")\n",
    "        \n",
    "        os.system(\"rm tmp.bed\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "b6db6875-a321-4bad-b9b5-95271c9e14c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n",
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) python/3.9-2022.05 => python/3.6-conda5.2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "d=\"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_type2_piRNA_08_20_2023/bed\"\n",
    "make_sense_bw(d, f'{d}/piRNA_bw', chrom_sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53878f66-a4e9-4937-b036-7c609d897985",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_anti_bw(direc, bwdir, chrom_sizes) : \n",
    "    \n",
    "    if not os.path.exists(bwdir) : \n",
    "        os.mkdir(bwdir)\n",
    "        \n",
    "    files = glob.glob(f\"{direc}/*.bed.tsv\")\n",
    "    \n",
    "    for f in files :\n",
    "        name = os.path.basename(f).split(\".\")[0]\n",
    "        d = pd.read_csv(f, sep = \"\\t\")\n",
    "        d_sub = d.query('feature == \"antipiRNA\"').reset_index(drop = True)\n",
    "        reformat = d_sub[['chrom', 'start', 'end', 'seq', 'count_total_norm']]\n",
    "        reformat.to_csv(\"tmp.bed\", sep = \"\\t\", header = False, index = False)\n",
    "\n",
    "        os.system(f\"bash ./bed2bw.sh tmp.bed {bwdir}/{name}.antipiRNA.bw {chrom_sizes} antipiRNA\")\n",
    "        \n",
    "        os.system(\"rm tmp.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58fdd1f5-b9f1-4542-aa29-4944600455ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "d=\"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_normalize_miRNA_07_31_2023/bed\"\n",
    "make_anti_bw(d, f'{d}/antipiRNA_raw_bw', chrom_sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7eaf36fd-1c0f-4a06-ba49-54138f30ef74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge bigWigs to collapse replicates\n",
    "\n",
    "def merge_bw(bws, chrom_sizes, name, outdir) :\n",
    "    \n",
    "    if not os.path.exists(outdir) : \n",
    "        os.mkdir(outdir)\n",
    "     \n",
    "    cmd = (f\"\"\"sh /fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/merge_bw.sh {chrom_sizes} {name} \"{' '.join(bws)}\" {outdir} {len(bws)}\"\"\")\n",
    "    os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14b34c4-ae6c-4ad3-897b-27c85f2b7481",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = \"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_normalize_miRNA_07_31_2023/bed/antipiRNA_raw_bw\"\n",
    "out = './merged_bw_0817_antiraw'\n",
    "merge_bw(bws = [os.path.join(d, 'parn_YD1.antipiRNA.bw'), \n",
    "                os.path.join(d, 'parn_YD2.antipiRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'parn_YD.antiraw', \n",
    "        outdir = out)\n",
    "\n",
    "merge_bw(bws = [os.path.join(d, 'N2_YD1.antipiRNA.bw'), \n",
    "                os.path.join(d, 'N2_YD2.antipiRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'N2_YD.antiraw', \n",
    "        outdir = out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a623e6fa-d793-45b2-bb20-1b7fab4f5686",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = \"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/results_07102023/bed/piRNA_bw\"\n",
    "out = './merged_bw_0816'\n",
    "merge_bw(bws = [os.path.join(d, 'parn_YD1.piRNA.bw'), \n",
    "                os.path.join(d, 'parn_YD2.piRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'parn_YD.piRNA', \n",
    "        outdir = out)\n",
    "\n",
    "merge_bw(bws = [os.path.join(d, 'N2_YD1.piRNA.bw'), \n",
    "                os.path.join(d, 'N2_YD2.piRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'N2_YD.piRNA', \n",
    "        outdir = out)\n",
    "\n",
    "merge_bw(bws = [os.path.join(d, 'parn1_w03del_1_S15_R1_001.piRNA.bw'), \n",
    "                os.path.join(d, 'parn1_w03del_2_S16_R1_001.piRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'parn1w03.piRNA', \n",
    "        outdir = out)\n",
    "\n",
    "d = \"/fs/ess/PCON0160/ben/projects/antisense_piRNA_v2/antipiRNA_bw_miRNA_norm\"\n",
    "merge_bw(bws = [os.path.join(d, 'parn_YD1.antipiRNA.bw'), \n",
    "                os.path.join(d, 'parn_YD2.antipiRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'parn_YD.antipiRNA', \n",
    "        outdir = out)\n",
    "\n",
    "merge_bw(bws = [os.path.join(d, 'N2_YD1.antipiRNA.bw'), \n",
    "                os.path.join(d, 'N2_YD1.antipiRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'N2_YD.antipiRNA', \n",
    "        outdir = out)\n",
    "\n",
    "merge_bw(bws = [os.path.join(d, 'parn1_w03del_1_S15_R1_001.antipiRNA.bw'), \n",
    "                os.path.join(d, 'parn1_w03del_2_S16_R1_001.antipiRNA.bw')], \n",
    "        chrom_sizes = chrom_sizes, \n",
    "        name = 'parn1w03.antipiRNA', \n",
    "        outdir = out)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536e341d-d4ce-4cab-bb7e-bbd8e868530f",
   "metadata": {},
   "source": [
    "# 4. calculate length distribution for piRNAs and antisense "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "37d0b602-fde1-40ba-a636-92dec7506e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def length_distribution(df) : \n",
    "    \n",
    "    df['first'] = df['seq'].str[0]\n",
    "    df['length'] = df['seq'].str.len()\n",
    "    \n",
    "    df_grouped = df.groupby(['first', 'length', 'feature', 'sample']).agg( {'seq':'count', 'count':'sum'}).reset_index()\n",
    "    return df_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "0ad5afca-f589-45ae-aa4f-c2b27b364bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "length_distribution_filt = length_distribution(filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "b74e9f78-a5a9-4a2d-a601-977bf9fb7cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "length_distribution_raw = length_distribution(piRNA_master_antisense)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "dc7730b6-e63f-4ae0-9612-c63d2bf87a17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>first</th>\n",
       "      <th>length</th>\n",
       "      <th>feature</th>\n",
       "      <th>sample</th>\n",
       "      <th>seq</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>N2NOT_raw</td>\n",
       "      <td>1</td>\n",
       "      <td>0.242184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>N2PP_raw</td>\n",
       "      <td>5</td>\n",
       "      <td>1.368273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>N2TAP_raw</td>\n",
       "      <td>3</td>\n",
       "      <td>0.425658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>N2_YD1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.233215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>N2_YD2</td>\n",
       "      <td>13</td>\n",
       "      <td>0.375875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  first  length    feature     sample  seq     count\n",
       "0     A      15  antipiRNA  N2NOT_raw    1  0.242184\n",
       "1     A      15  antipiRNA   N2PP_raw    5  1.368273\n",
       "2     A      15  antipiRNA  N2TAP_raw    3  0.425658\n",
       "3     A      15  antipiRNA     N2_YD1    3  0.233215\n",
       "4     A      15  antipiRNA     N2_YD2   13  0.375875"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_distribution_raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "e5ce4059-9a97-4777-801d-539c662f340f",
   "metadata": {},
   "outputs": [],
   "source": [
    "length_distribution_raw.to_csv(f\"./length_distribution_unfiltered_piRNA.tsv\", sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "094813f6-f9b6-4138-8c41-966cba5bbdee",
   "metadata": {},
   "source": [
    "# 5. length distribution of sense piRNAs unfiltered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "630018d1-9382-45d6-bfae-5577078b9540",
   "metadata": {},
   "outputs": [],
   "source": [
    "def piRNA_lendist(files) : \n",
    "\n",
    "    for e,f in enumerate(files) : \n",
    "        print(os.path.basename(f).split(\".\")[0])\n",
    "        samples = []\n",
    "        seqlens = []\n",
    "        firsts = []\n",
    "        rpms = []\n",
    "        features = []\n",
    "        sample = os.path.basename(f).split(\".\")[0]\n",
    "        with open(f, 'r') as f : \n",
    "            for line in f : \n",
    "                if not line.startswith(\"chrom\") :\n",
    "                    info = line.strip().split(\"\\t\")\n",
    "                    if info[7] == 'piRNA' or info[7] == \"redefine_type2_piRNA\" : \n",
    "                        seq = str(info[3])\n",
    "                        seql = len(seq)\n",
    "                        first = seq[0]\n",
    "                        rpm = float(info[13])\n",
    "\n",
    "                        seqlens.append(seql)\n",
    "                        firsts.append(first)\n",
    "                        rpms.append(rpm)\n",
    "                        samples.append(sample)\n",
    "                        features.append(info[7])\n",
    "        f.close()\n",
    "\n",
    "        df = pd.DataFrame({\n",
    "            'seqlen':seqlens,\n",
    "            'firstnt':firsts,\n",
    "            'rpm':rpms,\n",
    "            'sample':samples,\n",
    "            'feature' : features\n",
    "        })  \n",
    "\n",
    "        if e == 0 : \n",
    "            res = df \n",
    "        else : \n",
    "            res = pd.concat([res, df], ignore_index = True)\n",
    "\n",
    "    #res_grouped = res.groupby(['firstnt', 'seqlen', 'sample']).agg( {'rpm':'sum'}).reset_index()\n",
    "    \n",
    "    return res #res_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b130901d-9af8-4266-a936-c1aa76f61264",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob(\"./results_08_30_2023/bed/*.bed.tsv\") #+ glob.glob(\"./results_07_14_2023/bed/*.bed.tsv\") + glob.glob(\"./results_SNAP_ligation_07_24_2023/bed/*.bed.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d7bea6c7-5917-431a-9ffb-6a17cab251b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['./results_08_30_2023/bed/N2input_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/N2prgip_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/TW28prgip_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/TW29_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/TW28TAP_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/TW28PP_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/TW28input_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/N2NOT_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/N2TAP_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/N2PP_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv',\n",
       " './results_08_30_2023/bed/TW28NOT_raw.trimmed.uniq.x5p4.x3p0.uniq.xartifacts.xc.genome.aligned.v0.m1.combined.bed.tsv']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "079783b2-d965-4439-b015-79c21d13f2a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N2input_raw\n",
      "N2prgip_raw\n",
      "TW28prgip_raw\n",
      "TW29_raw\n",
      "TW28TAP_raw\n",
      "TW28PP_raw\n",
      "TW28input_raw\n",
      "N2NOT_raw\n",
      "N2TAP_raw\n",
      "N2PP_raw\n",
      "TW28NOT_raw\n"
     ]
    }
   ],
   "source": [
    "lendist = piRNA_lendist(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "192a092b-b93a-43dd-be69-8f04d3e8f747",
   "metadata": {},
   "outputs": [],
   "source": [
    "lendist.to_csv(\"./results_08_30_2023/piRNA_length_distribution.tsv\", sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f578bf4d-18ea-4b4e-8a2d-32ef69854f5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqlen</th>\n",
       "      <th>firstnt</th>\n",
       "      <th>rpm</th>\n",
       "      <th>sample</th>\n",
       "      <th>feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>0.879562</td>\n",
       "      <td>N2input_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>17.151459</td>\n",
       "      <td>N2input_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>28.585765</td>\n",
       "      <td>N2input_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>23</td>\n",
       "      <td>G</td>\n",
       "      <td>0.219891</td>\n",
       "      <td>N2input_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21</td>\n",
       "      <td>G</td>\n",
       "      <td>0.219891</td>\n",
       "      <td>N2input_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404692</th>\n",
       "      <td>19</td>\n",
       "      <td>A</td>\n",
       "      <td>1.054166</td>\n",
       "      <td>TW28NOT_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404693</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>0.527083</td>\n",
       "      <td>TW28NOT_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404694</th>\n",
       "      <td>18</td>\n",
       "      <td>A</td>\n",
       "      <td>2.108331</td>\n",
       "      <td>TW28NOT_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404695</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>1.581249</td>\n",
       "      <td>TW28NOT_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404696</th>\n",
       "      <td>24</td>\n",
       "      <td>G</td>\n",
       "      <td>0.527083</td>\n",
       "      <td>TW28NOT_raw</td>\n",
       "      <td>piRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>404697 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqlen firstnt        rpm       sample feature\n",
       "0           20       G   0.879562  N2input_raw   piRNA\n",
       "1           21       G  17.151459  N2input_raw   piRNA\n",
       "2           22       G  28.585765  N2input_raw   piRNA\n",
       "3           23       G   0.219891  N2input_raw   piRNA\n",
       "4           21       G   0.219891  N2input_raw   piRNA\n",
       "...        ...     ...        ...          ...     ...\n",
       "404692      19       A   1.054166  TW28NOT_raw   piRNA\n",
       "404693      17       A   0.527083  TW28NOT_raw   piRNA\n",
       "404694      18       A   2.108331  TW28NOT_raw   piRNA\n",
       "404695      17       A   1.581249  TW28NOT_raw   piRNA\n",
       "404696      24       G   0.527083  TW28NOT_raw   piRNA\n",
       "\n",
       "[404697 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lendist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2dcd943-b515-4642-b65c-27cb38528acd",
   "metadata": {},
   "source": [
    "# 6. make bed for specific examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "35eaf36d-0a08-48fe-8859-fed099a79523",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first make master table with all antisense / sense reads for piRNAs\n",
    "# combine tables\n",
    "# subset out specific piRNA of interest\n",
    "# export as bedfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bcc378e6-d973-4447-8095-fd37d89caf7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_sense_master(direc) : \n",
    "        \n",
    "    files = glob.glob(f\"{direc}/*.bed.tsv\")\n",
    "    \n",
    "    for e,f in enumerate(files) :\n",
    "        name = os.path.basename(f).split(\".\")[0]\n",
    "        print(name)\n",
    "        d = pd.read_csv(f, sep = \"\\t\")\n",
    "        d_sub = d.query('feature == \"piRNA\"').reset_index(drop = True).drop(columns = ['count'] )\n",
    "        d_sub = d_sub.rename(columns = {'count_total_norm' : 'count'})\n",
    "        d_sub['sample'] = name\n",
    "        if e == 0 : \n",
    "            res = d_sub\n",
    "        else : \n",
    "            res = pd.concat([d_sub, res], ignore_index = True)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9494a772-582d-4043-999f-aaf139810f06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N2_YD1\n",
      "N2_YD2\n",
      "parn1_prg1sg_2_S14_R1_001\n",
      "parn1_rrf1_redo_1_S21_R1_001\n",
      "parn1_drh3_2_S20_R1_001\n",
      "parn_YD2\n",
      "parn1ego1RNAi_2_S3_R1_001\n",
      "parn1ego1RNAi_1_S2_R1_001\n",
      "parn1_prg1DA_2_S12_R1_001\n",
      "parnOxiYD_2_R1\n",
      "parn1disl2_2_R1\n",
      "parn1-ekl1-RNAi-1_S4_R1_001\n",
      "parn1_prg1DA_1_S11_R1_001\n",
      "parn1_drh3RNAi_1_S17_R1_001\n",
      "parn1_prg1sg_1_S13_R1_001\n",
      "parnOxiYD_1_R1\n",
      "parn1disl2_1_R1\n",
      "parn1_drh3RNAi_2_S18_R1_001\n",
      "parn1_w03del_2_S16_R1_001\n",
      "parn1_rrf1_2_RENAMED_S19_R1_001\n",
      "parn1_drh3_RENAMED_1_S22_R1_001\n",
      "parn1rde3_2_R1\n",
      "parn1_oma1RNAi_2_S20_R1_001\n",
      "parn_YD1\n",
      "parn1-ekl1-RNAi-2_S5_R1_001\n",
      "parn1_w03del_1_S15_R1_001\n",
      "parn1rde3_1_R1\n",
      "parn1_oma1RNAi_1_S19_R1_001\n"
     ]
    }
   ],
   "source": [
    "sense1 = make_sense_master(\"./results_07102023/bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "979b1186-6412-4116-9fc0-a01ccad847f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TW28NOT_raw\n",
      "N2NOT_raw\n",
      "N2input_raw\n",
      "N2TAP_raw\n",
      "TW29_raw\n",
      "N2prgip_raw\n",
      "TW28TAP_raw\n",
      "TW28input_raw\n",
      "TW28PP_raw\n",
      "TW28prgip_raw\n",
      "N2PP_raw\n"
     ]
    }
   ],
   "source": [
    "sense2 = make_sense_master(\"./results_07_14_2023/bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0652c208-7312-436f-906f-7427c997ef93",
   "metadata": {},
   "outputs": [],
   "source": [
    "anti1 = pd.read_csv(\"./results_07102023/bed/antipiRNA_bed_master_filtered.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3c7da3ab-f572-44d5-bb31-233eb2486ad1",
   "metadata": {},
   "outputs": [],
   "source": [
    "anti2 = pd.read_csv(\"./results_07_14_2023/bed/antipiRNA_bed_master_filtered.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "29825abd-7dbb-4889-8a4a-06be5ee65700",
   "metadata": {},
   "outputs": [],
   "source": [
    "anti1 = anti1[['chrom', 'start', 'end', 'seq', 'feature', 'gene', 'sample', 'count']]\n",
    "anti2 = anti2[['chrom', 'start', 'end', 'seq', 'feature', 'gene', 'sample', 'count']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "94c1f395-f56c-430f-bf25-5a225adcaca2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sense1 = sense1[['chrom', 'start', 'end', 'seq', 'feature', 'gene', 'sample', 'count']]\n",
    "sense2 = sense2[['chrom', 'start', 'end', 'seq', 'feature', 'gene', 'sample', 'count']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ed832ede-ee5e-48ab-9c75-8803341abee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "master = pd.concat([sense1, anti1, sense2, anti2], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7c19f88b-4c54-4b32-beb3-0cd840758d03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>seq</th>\n",
       "      <th>feature</th>\n",
       "      <th>gene</th>\n",
       "      <th>sample</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>5330474</td>\n",
       "      <td>5330498</td>\n",
       "      <td>TCAACAATAAAAATTGAGGAGTAG</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>WBGene00170676,F48C1.10,21ur-12374</td>\n",
       "      <td>parn1_oma1RNAi_1_S19_R1_001</td>\n",
       "      <td>0.140525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>5330474</td>\n",
       "      <td>5330501</td>\n",
       "      <td>TCAACAATAAAAATTGAGGAGTAGATA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>WBGene00170676,F48C1.10,21ur-12374</td>\n",
       "      <td>parn1_oma1RNAi_1_S19_R1_001</td>\n",
       "      <td>0.070262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>5330474</td>\n",
       "      <td>5330503</td>\n",
       "      <td>TCAACAATAAAAATTGAGGAGTAGATAGA</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>WBGene00170676,F48C1.10,21ur-12374</td>\n",
       "      <td>parn1_oma1RNAi_1_S19_R1_001</td>\n",
       "      <td>0.035131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>5431491</td>\n",
       "      <td>5431508</td>\n",
       "      <td>TGCACCTCGTGAGAAGT</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>WBGene00170835,F27C1.14,21ur-11585</td>\n",
       "      <td>parn1_oma1RNAi_1_S19_R1_001</td>\n",
       "      <td>0.035131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>5486066</td>\n",
       "      <td>5486089</td>\n",
       "      <td>TATTTTGAATAAAATTTTCGGGG</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>WBGene00174770,C09D4.7,21ur-12380</td>\n",
       "      <td>parn1_oma1RNAi_1_S19_R1_001</td>\n",
       "      <td>0.070262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2507901</th>\n",
       "      <td>IV</td>\n",
       "      <td>16914647</td>\n",
       "      <td>16914666</td>\n",
       "      <td>ATACTTCTATGCATGACAA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00048497,Y116A8C.128,21ur-5307</td>\n",
       "      <td>N2PP_raw</td>\n",
       "      <td>0.273655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2507902</th>\n",
       "      <td>IV</td>\n",
       "      <td>16914649</td>\n",
       "      <td>16914666</td>\n",
       "      <td>ACTTCTATGCATGACAA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00048497,Y116A8C.128,21ur-5307</td>\n",
       "      <td>N2PP_raw</td>\n",
       "      <td>0.273655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2507903</th>\n",
       "      <td>IV</td>\n",
       "      <td>16981618</td>\n",
       "      <td>16981636</td>\n",
       "      <td>ACGTGAATTTGGTGAATA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00050153,Y116A8C.170,21ur-1526</td>\n",
       "      <td>N2PP_raw</td>\n",
       "      <td>0.273655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2507904</th>\n",
       "      <td>IV</td>\n",
       "      <td>17023991</td>\n",
       "      <td>17024010</td>\n",
       "      <td>ATATGGTAAAAAGTTGAAA</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00173668,Y116A8C.420,21ur-9954</td>\n",
       "      <td>N2PP_raw</td>\n",
       "      <td>0.273655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2507905</th>\n",
       "      <td>IV</td>\n",
       "      <td>17186499</td>\n",
       "      <td>17186516</td>\n",
       "      <td>GTACTGATTAGTTGAAG</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>WBGene00050608,C52D10.57,21ur-258</td>\n",
       "      <td>N2PP_raw</td>\n",
       "      <td>0.273655</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2507906 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end                            seq    feature  \\\n",
       "0           I   5330474   5330498       TCAACAATAAAAATTGAGGAGTAG      piRNA   \n",
       "1           I   5330474   5330501    TCAACAATAAAAATTGAGGAGTAGATA      piRNA   \n",
       "2           I   5330474   5330503  TCAACAATAAAAATTGAGGAGTAGATAGA      piRNA   \n",
       "3           I   5431491   5431508              TGCACCTCGTGAGAAGT      piRNA   \n",
       "4           I   5486066   5486089        TATTTTGAATAAAATTTTCGGGG      piRNA   \n",
       "...       ...       ...       ...                            ...        ...   \n",
       "2507901    IV  16914647  16914666            ATACTTCTATGCATGACAA  antipiRNA   \n",
       "2507902    IV  16914649  16914666              ACTTCTATGCATGACAA  antipiRNA   \n",
       "2507903    IV  16981618  16981636             ACGTGAATTTGGTGAATA  antipiRNA   \n",
       "2507904    IV  17023991  17024010            ATATGGTAAAAAGTTGAAA  antipiRNA   \n",
       "2507905    IV  17186499  17186516              GTACTGATTAGTTGAAG  antipiRNA   \n",
       "\n",
       "                                         gene                       sample  \\\n",
       "0          WBGene00170676,F48C1.10,21ur-12374  parn1_oma1RNAi_1_S19_R1_001   \n",
       "1          WBGene00170676,F48C1.10,21ur-12374  parn1_oma1RNAi_1_S19_R1_001   \n",
       "2          WBGene00170676,F48C1.10,21ur-12374  parn1_oma1RNAi_1_S19_R1_001   \n",
       "3          WBGene00170835,F27C1.14,21ur-11585  parn1_oma1RNAi_1_S19_R1_001   \n",
       "4           WBGene00174770,C09D4.7,21ur-12380  parn1_oma1RNAi_1_S19_R1_001   \n",
       "...                                       ...                          ...   \n",
       "2507901  WBGene00048497,Y116A8C.128,21ur-5307                     N2PP_raw   \n",
       "2507902  WBGene00048497,Y116A8C.128,21ur-5307                     N2PP_raw   \n",
       "2507903  WBGene00050153,Y116A8C.170,21ur-1526                     N2PP_raw   \n",
       "2507904  WBGene00173668,Y116A8C.420,21ur-9954                     N2PP_raw   \n",
       "2507905     WBGene00050608,C52D10.57,21ur-258                     N2PP_raw   \n",
       "\n",
       "            count  \n",
       "0        0.140525  \n",
       "1        0.070262  \n",
       "2        0.035131  \n",
       "3        0.035131  \n",
       "4        0.070262  \n",
       "...           ...  \n",
       "2507901  0.273655  \n",
       "2507902  0.273655  \n",
       "2507903  0.273655  \n",
       "2507904  0.273655  \n",
       "2507905  0.273655  \n",
       "\n",
       "[2507906 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1ba4e5ca-369f-4598-98eb-dbc0b8a692f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#colnames = c(\"gene\", \"start\", \"end\", \"seq\", \"rpm\", \"strand\", \"sample\")\n",
    "\n",
    "gene = 'WBGene00048837'\n",
    "samples = ['N2_YD1', 'N2_YD2', 'parn_YD1', 'parn_YD2']\n",
    "sub = master.query('gene.str.contains(@gene) and sample.isin(@samples)')\n",
    "sub\n",
    "sub = sub[['chrom', 'start', 'end', 'seq', 'count', 'feature', 'sample']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "91170f56-171d-44e1-9aab-5f794476be7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>seq</th>\n",
       "      <th>count</th>\n",
       "      <th>feature</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>381545</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230629</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGACATT</td>\n",
       "      <td>0.055672</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381546</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230630</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGACAT</td>\n",
       "      <td>0.389701</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381547</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230631</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGACA</td>\n",
       "      <td>0.723731</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381548</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230632</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGAC</td>\n",
       "      <td>1.503134</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381549</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230633</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGA</td>\n",
       "      <td>0.612388</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381550</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230634</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATG</td>\n",
       "      <td>0.055672</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381551</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230635</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAAT</td>\n",
       "      <td>0.111343</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381552</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230637</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACA</td>\n",
       "      <td>0.111343</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381553</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230639</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTA</td>\n",
       "      <td>0.055672</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381554</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230640</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGT</td>\n",
       "      <td>0.222686</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874415</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230629</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGACATT</td>\n",
       "      <td>0.243103</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874416</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230630</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGACAT</td>\n",
       "      <td>0.291723</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874417</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230631</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGACA</td>\n",
       "      <td>0.777929</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874418</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230632</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGAC</td>\n",
       "      <td>0.972411</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874419</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230633</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAATGA</td>\n",
       "      <td>0.583447</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874420</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230635</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACAAT</td>\n",
       "      <td>0.097241</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874421</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230638</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTAC</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874422</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230640</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGT</td>\n",
       "      <td>0.194482</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1874423</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230641</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGG</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081332</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230637</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACA</td>\n",
       "      <td>13.638879</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081333</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230638</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTAC</td>\n",
       "      <td>0.187937</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081334</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230639</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTA</td>\n",
       "      <td>0.080545</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081335</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230640</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGT</td>\n",
       "      <td>0.214785</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2081336</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230641</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGG</td>\n",
       "      <td>0.026848</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2110045</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230637</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTACA</td>\n",
       "      <td>16.636005</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2110046</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230638</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGTAC</td>\n",
       "      <td>0.077738</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2110047</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230640</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGGT</td>\n",
       "      <td>0.544168</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2110048</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230641</td>\n",
       "      <td>15230658</td>\n",
       "      <td>TCGAATGCATTCTTAGG</td>\n",
       "      <td>0.077738</td>\n",
       "      <td>piRNA</td>\n",
       "      <td>N2_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2167220</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230629</td>\n",
       "      <td>15230647</td>\n",
       "      <td>AATGTCATTGTACCTAAG</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2167221</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230629</td>\n",
       "      <td>15230648</td>\n",
       "      <td>AATGTCATTGTACCTAAGA</td>\n",
       "      <td>0.097241</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2167222</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230630</td>\n",
       "      <td>15230648</td>\n",
       "      <td>ATGTCATTGTACCTAAGA</td>\n",
       "      <td>0.145862</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2167223</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230630</td>\n",
       "      <td>15230649</td>\n",
       "      <td>ATGTCATTGTACCTAAGAA</td>\n",
       "      <td>0.632067</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>parn_YD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2194002</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230629</td>\n",
       "      <td>15230648</td>\n",
       "      <td>AATGTCATTGTACCTAAGA</td>\n",
       "      <td>0.055672</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2194003</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230630</td>\n",
       "      <td>15230648</td>\n",
       "      <td>ATGTCATTGTACCTAAGA</td>\n",
       "      <td>0.222686</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2194004</th>\n",
       "      <td>IV</td>\n",
       "      <td>15230630</td>\n",
       "      <td>15230649</td>\n",
       "      <td>ATGTCATTGTACCTAAGAA</td>\n",
       "      <td>0.668059</td>\n",
       "      <td>antipiRNA</td>\n",
       "      <td>parn_YD1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end                            seq      count  \\\n",
       "381545     IV  15230629  15230658  TCGAATGCATTCTTAGGTACAATGACATT   0.055672   \n",
       "381546     IV  15230630  15230658   TCGAATGCATTCTTAGGTACAATGACAT   0.389701   \n",
       "381547     IV  15230631  15230658    TCGAATGCATTCTTAGGTACAATGACA   0.723731   \n",
       "381548     IV  15230632  15230658     TCGAATGCATTCTTAGGTACAATGAC   1.503134   \n",
       "381549     IV  15230633  15230658      TCGAATGCATTCTTAGGTACAATGA   0.612388   \n",
       "381550     IV  15230634  15230658       TCGAATGCATTCTTAGGTACAATG   0.055672   \n",
       "381551     IV  15230635  15230658        TCGAATGCATTCTTAGGTACAAT   0.111343   \n",
       "381552     IV  15230637  15230658          TCGAATGCATTCTTAGGTACA   0.111343   \n",
       "381553     IV  15230639  15230658            TCGAATGCATTCTTAGGTA   0.055672   \n",
       "381554     IV  15230640  15230658             TCGAATGCATTCTTAGGT   0.222686   \n",
       "1874415    IV  15230629  15230658  TCGAATGCATTCTTAGGTACAATGACATT   0.243103   \n",
       "1874416    IV  15230630  15230658   TCGAATGCATTCTTAGGTACAATGACAT   0.291723   \n",
       "1874417    IV  15230631  15230658    TCGAATGCATTCTTAGGTACAATGACA   0.777929   \n",
       "1874418    IV  15230632  15230658     TCGAATGCATTCTTAGGTACAATGAC   0.972411   \n",
       "1874419    IV  15230633  15230658      TCGAATGCATTCTTAGGTACAATGA   0.583447   \n",
       "1874420    IV  15230635  15230658        TCGAATGCATTCTTAGGTACAAT   0.097241   \n",
       "1874421    IV  15230638  15230658           TCGAATGCATTCTTAGGTAC   0.048621   \n",
       "1874422    IV  15230640  15230658             TCGAATGCATTCTTAGGT   0.194482   \n",
       "1874423    IV  15230641  15230658              TCGAATGCATTCTTAGG   0.048621   \n",
       "2081332    IV  15230637  15230658          TCGAATGCATTCTTAGGTACA  13.638879   \n",
       "2081333    IV  15230638  15230658           TCGAATGCATTCTTAGGTAC   0.187937   \n",
       "2081334    IV  15230639  15230658            TCGAATGCATTCTTAGGTA   0.080545   \n",
       "2081335    IV  15230640  15230658             TCGAATGCATTCTTAGGT   0.214785   \n",
       "2081336    IV  15230641  15230658              TCGAATGCATTCTTAGG   0.026848   \n",
       "2110045    IV  15230637  15230658          TCGAATGCATTCTTAGGTACA  16.636005   \n",
       "2110046    IV  15230638  15230658           TCGAATGCATTCTTAGGTAC   0.077738   \n",
       "2110047    IV  15230640  15230658             TCGAATGCATTCTTAGGT   0.544168   \n",
       "2110048    IV  15230641  15230658              TCGAATGCATTCTTAGG   0.077738   \n",
       "2167220    IV  15230629  15230647             AATGTCATTGTACCTAAG   0.048621   \n",
       "2167221    IV  15230629  15230648            AATGTCATTGTACCTAAGA   0.097241   \n",
       "2167222    IV  15230630  15230648             ATGTCATTGTACCTAAGA   0.145862   \n",
       "2167223    IV  15230630  15230649            ATGTCATTGTACCTAAGAA   0.632067   \n",
       "2194002    IV  15230629  15230648            AATGTCATTGTACCTAAGA   0.055672   \n",
       "2194003    IV  15230630  15230648             ATGTCATTGTACCTAAGA   0.222686   \n",
       "2194004    IV  15230630  15230649            ATGTCATTGTACCTAAGAA   0.668059   \n",
       "\n",
       "           feature    sample  \n",
       "381545       piRNA  parn_YD1  \n",
       "381546       piRNA  parn_YD1  \n",
       "381547       piRNA  parn_YD1  \n",
       "381548       piRNA  parn_YD1  \n",
       "381549       piRNA  parn_YD1  \n",
       "381550       piRNA  parn_YD1  \n",
       "381551       piRNA  parn_YD1  \n",
       "381552       piRNA  parn_YD1  \n",
       "381553       piRNA  parn_YD1  \n",
       "381554       piRNA  parn_YD1  \n",
       "1874415      piRNA  parn_YD2  \n",
       "1874416      piRNA  parn_YD2  \n",
       "1874417      piRNA  parn_YD2  \n",
       "1874418      piRNA  parn_YD2  \n",
       "1874419      piRNA  parn_YD2  \n",
       "1874420      piRNA  parn_YD2  \n",
       "1874421      piRNA  parn_YD2  \n",
       "1874422      piRNA  parn_YD2  \n",
       "1874423      piRNA  parn_YD2  \n",
       "2081332      piRNA    N2_YD2  \n",
       "2081333      piRNA    N2_YD2  \n",
       "2081334      piRNA    N2_YD2  \n",
       "2081335      piRNA    N2_YD2  \n",
       "2081336      piRNA    N2_YD2  \n",
       "2110045      piRNA    N2_YD1  \n",
       "2110046      piRNA    N2_YD1  \n",
       "2110047      piRNA    N2_YD1  \n",
       "2110048      piRNA    N2_YD1  \n",
       "2167220  antipiRNA  parn_YD2  \n",
       "2167221  antipiRNA  parn_YD2  \n",
       "2167222  antipiRNA  parn_YD2  \n",
       "2167223  antipiRNA  parn_YD2  \n",
       "2194002  antipiRNA  parn_YD1  \n",
       "2194003  antipiRNA  parn_YD1  \n",
       "2194004  antipiRNA  parn_YD1  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "15ac3d01-9550-480b-b4c3-015708e8135b",
   "metadata": {},
   "outputs": [],
   "source": [
    "specific_example_dir = './specific_examples'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a1c11499-ff7d-493b-87b3-56a249f051af",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(specific_example_dir) : \n",
    "    os.mkdir(specific_example_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "53b72716-346e-44a3-b32e-04cad3d4e019",
   "metadata": {},
   "outputs": [],
   "source": [
    "sub.to_csv(f\"./specific_examples/{gene}.bed\", sep = '\\t', header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c078328-3a71-4cd8-8b88-2da10ca51926",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9 (Conda 2022.05) [python/3.9-2022.05]",
   "language": "python",
   "name": "python39_202205"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
